### PR TITLE
Python: Several standard library models

### PR DESCRIPTION
**Pull request by yoff**
This PR is merging the long-lived branch containing the modelling work to stop extracting the standard library by default.

#### All query authors

- [ ] A change note is added if necessary. See [the documentation](https://github.com/github/codeql/blob/main/docs/change-notes.md) in this repository.
- [x] QL tests are added if necessary. See [Testing custom queries](https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/testing-custom-queries) in the GitHub documentation.

#### Internal query authors only

- [ ] Autofixes generated based on these changes are valid, only needed if this PR makes significant changes to `.ql`, `.qll`, or `.qhelp` files. See [the documentation](https://github.com/github/codeql-team/blob/main/docs/best-practices/validating-autofix-for-query-changes.md) (internal access required).
- [x] Changes are validated [at scale](https://github.com/github/codeql-dca/) (internal access required).

### CODE REVIEW

File: python/ql/lib/change-notes/2024-09-24-std-lib-models.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* Added several models of standard library functions and classes. This should make the analysis much more complete in cases where the standard library is not extracted.
```
Comment (RasmusWL) [inline python/ql/lib/change-notes/2024-09-24-std-lib-models.md:4]: This change-note feels a bit vague, and to fully understand it you need to know that we're planning to not extract the standard library in the future. (which doesn't seem reasonable that all end-users will be aware of).

How about this instead?

```suggestion
* Added several models of standard library functions and classes, in anticipation of no longer extracting the standard library in a future release.
```

**Review by RasmusWL — CHANGES_REQUESTED** [review]
I think I already approved all these changes as part of other chains of PRs, right?

In that case, I only have one nitpick: With this PR we have two files for modeling stdlib with MaD (https://github.com/github/codeql/blob/main/python/ql/lib/semmle/python/frameworks/Stdlib.model.yml) so we should consolidate to only using one location. Personally I feel the obvious place to put modeling of a new library would be `ql/lib/semmle/python/frameworks/<package>.model.yml` (such as the one for [Asyncpg](https://github.com/github/codeql/blob/main/python/ql/lib/semmle/python/frameworks/Asyncpg.model.yml)). So unless you strongly object, would you mind moving the new models to `python/ql/lib/semmle/python/frameworks/Stdlib.model.yml`? :pray:

**Comment by yoff** [comment]
> With this PR we have two files for modeling stdlib with MaD

Nice. I had totally missed that we had started using MaD for StdLib. In that case I will put the new models there too. I will also adjust https://github.com/github/codeql/pull/17565.

**Review by tausbn — APPROVED** [review]

File: python/ql/lib/semmle/python/frameworks/Stdlib.model.yml [hunk 2]
```diff
@@ -21,9 +21,161 @@ extensions:
...
+
   - addsTo:
       pack: codeql/python-all
       extensible: summaryModel
     data:
-      - ['argparse.ArgumentParser', 'Member[parse_args,parse_known_args]', 'Argument[0,args:]', 'ReturnValue', 'taint']
+      # See https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser
+      - ["argparse.ArgumentParser", "Member[_parse_known_args,_read_args_from_files]", "Argument[0,arg_strings:]", "ReturnValue", "taint"]
       # note: taint of attribute lookups is handled in QL
+      - ["argparse.ArgumentParser", "Member[parse_args,parse_known_args]", "Argument[0,args:]", "ReturnValue", "taint"]
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/frameworks/Stdlib.model.yml:38]: I don't think my original comment was super clear, so here's a suggestion to clean that up a bit :sweat_smile: (found QL code [here](https://github.com/github/codeql/blob/7483075b7eeb577c5bedb43a9ce3b6a37e427dc2/python/ql/lib/semmle/python/frameworks/Stdlib.qll#L5005-L5020) for reference)

```suggestion
      # See https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser
      # note: taint flow for attribute lookups on `argparse.ArgumentParser` is handled in QL
      - ["argparse.ArgumentParser", "Member[_parse_known_args,_read_args_from_files]", "Argument[0,arg_strings:]", "ReturnValue", "taint"]
      - ["argparse.ArgumentParser", "Member[parse_args,parse_known_args]", "Argument[0,args:]", "ReturnValue", "taint"]
```

Comment (yoff) [reply python/ql/lib/semmle/python/frameworks/Stdlib.model.yml]: Ah, thanks, I was a bit puzzled by that comment; it makes much more sense now :-)

**Review by RasmusWL — APPROVED** [review]

**Pull request merged**